### PR TITLE
Tabs: remove custom logic

### PR DESCRIPTION
**Pull request by ciampo**
## What?

Related to #66011

Remove most custom logic in the `Tabs` component

## Why?

For a detailed explanation of the logic being removed, refer to the behaviors 1 to 4 listed in #66011. Behavior 6 is being kept for the time being, while the logic is incorporated directly in `ariakit` (see https://github.com/ariakit/ariakit/issues/4213).

- Some logic felt arbitrary, and it was agreed that if such a niche behavior would pop up, it should be addressed by the tab consumer;
- Some logic felt unnecessary (either fixed by ariakit in the meantime, or just extra)

## How?

Custom logic being removed (from #66011):

> * [Initial tab selection](https://github.com/WordPress/gutenberg/blob/609a3c8d38a0608b41b1827ec4a8c1f812487ef2/packages/components/src/tabs/index.tsx#L72-L109):
>   
>   * if a `defaultTabId` is specified, `Tabs` will select an initial tab only if there is a tab matching `defaultTabId`. If there isn't a matching tab, `Tabs` won't have an initially selected tab. If a tab with a matching id is added lazily, `Tabs` will select it.
>   * the initial tab is also used as a fallback when the currently selected tab can't be found anymore. If the initial tab also can't be found, the `Tabs` component won't have any selected tab;
>   * This bunch of logic only applied in uncontrolled mode — the idea being that a consumer controlling the component is in charge of handling these edge cases;
> * [The currently selected tab becomes disabled](https://github.com/WordPress/gutenberg/blob/609a3c8d38a0608b41b1827ec4a8c1f812487ef2/packages/components/src/tabs/index.tsx#L111-L141):
>   
>   * the main assumption is that a disabled tab cannot stay as the selected tab;
>   * In controlled mode, `Tabs` assumes that the consumer of the component is in charge of the situation, and therefore it un-selects the previously active, now disabled tab;
>   * In uncontrolled more, if the currently selected tab becomes disabled, `Tabs` falls back to the `defaultTabId` if possible. Otherwise, it selects the first enabled tab (if there is one).
> * In controlled mode, [`Tabs` will reset the active tab id (ie. no tabs are active) if a tab associated to the currently selected ID can't be found](https://github.com/WordPress/gutenberg/blob/609a3c8d38a0608b41b1827ec4a8c1f812487ef2/packages/components/src/tabs/index.tsx#L143-L158)
> * [If there is no active tab, fallback to place focus on the first enabled tab, so there is always an active element](https://github.com/WordPress/gutenberg/blob/609a3c8d38a0608b41b1827ec4a8c1f812487ef2/packages/components/src/tabs/index.tsx#L160-L166)

Process:

- Remove one hook at a time
- Fix tests
- Repeat
- Move custom logic in consumer-land to handle special cases

## Testing Instructions

- Look at the unit test changes, understand if the changes feel reasonable;
- Smoke test tabs usage in Storybook and in the editor.
  - In particular, [this usage](https://github.com/WordPress/gutenberg/pull/66097#discussion_r1801253430) should continue to work as expected

Specific behaviours to check:

- if no `selectedId` or `defaultSelectedId` props are passed and refer to an existing tab id, `Tabs` will automatically select the first non-disabled tab and shows related tab-panel;
- if `selectedId` or `defaultSelectedId` props are passed but refer to a tab id that doesn't exist, `Tabs` won't show any tabs as selected and won't show any tabpanels;
- if the `selectedId` prop is passed but refers to a tab id that doesn't exist, and the tab is later added to the react tree, `Tabs` will correctly select that tab; this won't work with `defaultSelectedId`, since that prop only works for the initial render;
- if `selectedId` or `defaultSelectedId` props have a value of `null`, no tabs will be selected and no tabpanels will be shown, and the tablist itself will become tabbable; when focus in on the tablist, pressing arrow keys will select tabs;
- if `selectedId` or `defaultSelectedId` props are passed and refer to a disabled tab, that tab will be selected anyway (and its associated tabpanel will be shown);
- if the currently selected tab becomes disabled, it continues to say selected (and its associated tabpanel continues to show);
- - if the currently selected tab gets removed, no tabs are selected (and no tab panels are shown);
- if for any reason there is no active tab id (for example, as a result of clearing the selected tab id, or for removing the previously active tab from the DOM), all tabs in the tablist should become tabbable so that they can capture keyboard focus. As soon as a tab receives focus, all tabs will go back to default behaviour.

Note: there is [a toolbar e2e test that is expected to fail](https://github.com/WordPress/gutenberg/pull/65907#issuecomment-2406408450), since this PR is currently based off #65907.

For added context, here is also a (potentially incomplete) history of recent changes to the `Tabs` component, specifically to how it handles initial selection, disabled tabs, removed tabs, and general focus handling:

- https://github.com/WordPress/gutenberg/pull/55287
- https://github.com/WordPress/gutenberg/pull/56658
- https://github.com/WordPress/gutenberg/pull/57696
- https://github.com/WordPress/gutenberg/pull/58625
- https://github.com/WordPress/gutenberg/pull/60681

### CODE REVIEW

File: packages/block-editor/src/components/inserter/category-tabs/index.js [hunk 1]
```diff
@@ -31,11 +32,23 @@ function CategoryTabs( {
...
+	const firstTabId = categories?.[ 0 ]?.name;
+	useEffect( () => {
+		// If there is no active tab, make the first tab the active tab, so that
+		// when focus is moved to the tablist, the first tab will be focused
+		// despite not being selected
+		if ( selectedTabId === null && ! activeTabId && firstTabId ) {
+			setActiveId( firstTabId );
+		}
+	}, [ selectedTabId, activeTabId, firstTabId, setActiveId ] );
+
```
Comment (ciampo) [inline packages/block-editor/src/components/inserter/category-tabs/index.js:46]: The custom logic added in https://github.com/WordPress/gutenberg/pull/60681 is now applied directly where `Tabs` is consumed. The result should be the same. (cc @jeryj )

File: packages/components/src/tabs/tablist.tsx [hunk 2]
```diff
@@ -112,9 +112,13 @@ export const TabList = forwardRef<
 		<Ariakit.TabList
 			ref={ refs }
 			store={ store }
-			render={ <TabListWrapper /> }
+			render={ ( props ) => (
+				<TabListWrapper
+					{ ...props }
+					tabIndex={ props.tabIndex ?? -1 }
+				/>
+			) }
```
Comment (ciampo) [inline packages/components/src/tabs/tablist.tsx]: This is a bug fix that can be extracted and merged on `trunk`

File: packages/components/src/tabs/types.ts [hunk 3]
```diff
@@ -65,6 +65,27 @@ export type TabsProps = {
...
+	 * The tab id that should be active by default when the composite widget is
+	 * rendered. If `null`, the tablist element itself will have focus
+	 * and users will be able to navigate to it using arrow keys. If `undefined`,
+	 * the first enabled item will be focused.
+	 */
+	defaultActiveTabId?: Ariakit.CompositeStoreProps[ 'defaultActiveId' ];
+	/**
+	 * A callback that gets called when the `activeId` state changes.
+	 */
+	onActiveTabIdChange?: Ariakit.TabStoreProps[ 'setActiveId' ];
```
Comment (ciampo) [inline packages/components/src/tabs/types.ts]: The addition of these props, and their usage in the patterns inserter can be extracted to a separate PR

**Comment by ciampo** [comment]
I think this PR is ready for a first round of review / smoke testing, at least to get a feeling about whether we're moving in the right direction.

File: packages/block-editor/src/components/inserter/category-tabs/index.js (hunk 1 above)
Comment (jeryj) [reply packages/block-editor/src/components/inserter/category-tabs/index.js:46]: I tested on the Patterns tab of the inserter, and confirmed it focuses the first tab without activating it. It works the same as on trunk. Thanks!

File: packages/components/src/tabs/tablist.tsx (hunk 2 above)
Comment (ciampo) [reply packages/components/src/tabs/tablist.tsx]: Fix extracted in https://github.com/WordPress/gutenberg/pull/66209

File: packages/components/src/tabs/types.ts (hunk 3 above)
Comment (ciampo) [reply packages/components/src/tabs/types.ts]: Added in https://github.com/WordPress/gutenberg/pull/66223

**Comment by ciampo** [comment]
Excuses for the many pings — not sure why GitHub decided to add as many reviewers on my behalf as the result of a rebase + target branch change. Please feel free to ignore this PR if you're _currently_ listed as a reviewer

File: packages/components/src/tabs/index.tsx [hunk 4]
```diff
@@ -40,147 +50,41 @@ export const Tabs = Object.assign(
...
-				!! selectedTabId &&
-				! selectedTab
-			) {
-				setSelectedId( null );
-			}
-		}, [ isControlled, selectedTab, selectedTabId, setSelectedId ] );
+		const { items, activeId } = Ariakit.useStoreState( store );
+		const { setActiveId } = store;
 
 		useEffect( () => {
```
Comment (ciampo) [inline packages/components/src/tabs/index.tsx:87]: The behaviour applied by this custom hook is currently being in https://github.com/ariakit/ariakit/issues/4213 , to understand whether it will be even incorporated into ariakit, or if it's even a reasonable fix to implement on our side in the first place.

It will be kept around for the time being.

**Comment by tyxla** [comment]
@ciampo is this one ready for review after the related PRs landed? If so, would it be possible to add some manual testing instructions?

**Comment by ciampo** [comment]
Yes, this PR is ready to review. I'll prepare some additional manual testing instructions specifically to the behavioural changes that happen when removing the custom logic, and ping you when they're ready.

File: packages/components/src/tabs/index.tsx [hunk 5]
```diff
@@ -200,7 +104,7 @@ export const Tabs = Object.assign(
 					setActiveId( focusedElement.id );
```
Comment (ciampo) [inline packages/components/src/tabs/index.tsx:104]: The behaviour applied by this last custom hook is currently being discussed in https://github.com/ariakit/ariakit/issues/4213 , to understand whether it will be even incorporated into ariakit, or if it's even a reasonable fix to implement on our side in the first place.

It will be kept around for the time being.

File: packages/components/src/tabs/test/index.tsx [hunk 6]
```diff
@@ -191,6 +195,8 @@ describe( 'Tabs', () => {
 		it( 'should focus on the related TabPanel when pressing the Tab key', async () => {
 			await render( <UncontrolledTabs tabs={ TABS } /> );
 
+			expect( await getSelectedTab() ).toHaveTextContent( 'Alpha' );
```
Comment (ciampo) [inline packages/components/src/tabs/test/index.tsx:198]: All of these extra, initial assertions were added to increase test stability — basically, we are waiting for ariakit to initialise and we do that by waiting for the correct tab to be selected.

File: packages/components/src/tabs/test/index.tsx [hunk 7]
```diff
@@ -403,14 +430,14 @@ describe( 'Tabs', () => {
...
+			expect( await getSelectedTab() ).not.toHaveFocus();
+
+			// onSelect gets called on the initial render.
 			expect( mockOnSelect ).toHaveBeenCalledTimes( 1 );
 			expect( mockOnSelect ).toHaveBeenLastCalledWith( 'alpha' );
 
-			// Tab to focus the tablist. Make sure alpha is focused.
-			expect( await getSelectedTab() ).toHaveTextContent( 'Alpha' );
-			expect( await getSelectedTab() ).not.toHaveFocus();
```
Comment (ciampo) [inline packages/components/src/tabs/test/index.tsx:413]: no behavioural change here, just moved assertions around to ensure better test stability

File: packages/components/src/tabs/test/index.tsx [hunk 8]
```diff
@@ -502,16 +529,17 @@ describe( 'Tabs', () => {
...
+			expect( await getSelectedTab() ).toHaveTextContent( 'Alpha' );
+			expect( await getSelectedTab() ).not.toHaveFocus();
+
+			// onSelect gets called on the initial render.
 			expect( mockOnSelect ).toHaveBeenCalledTimes( 1 );
 			expect( mockOnSelect ).toHaveBeenLastCalledWith( 'alpha' );
 
 			// Tab to focus the tablist. Make sure Alpha is focused.
-			expect( await getSelectedTab() ).toHaveTextContent( 'Alpha' );
-			expect( await getSelectedTab() ).not.toHaveFocus();
```
Comment (ciampo) [inline packages/components/src/tabs/test/index.tsx:512]: no behavioural change here, just moved assertions around to ensure better test stability.

File: packages/components/src/tabs/test/index.tsx [hunk 9]
```diff
@@ -635,39 +668,20 @@ describe( 'Tabs', () => {
 					await screen.findByRole( 'tabpanel', { name: 'Alpha' } )
 				).toBeInTheDocument();
 			} );
-			it( 'should fall back to first enabled tab if the active tab is removed', async () => {
+			it( 'should not have a selected tab if the currently selected tab is removed', async () => {
```
Comment (ciampo) [inline packages/components/src/tabs/test/index.tsx:671]: This test has changed to test the new behaviour

File: packages/components/src/tabs/test/index.tsx [hunk 10]
```diff
@@ -715,14 +729,16 @@ describe( 'Tabs', () => {
...
+
 				await rerender(
 					<UncontrolledTabs tabs={ TABS } defaultTabId="alpha" />
 				);
 
 				expect( await getSelectedTab() ).toHaveTextContent( 'Beta' );
 			} );
 
-			it( 'should fall back to the tab associated to `defaultTabId` if the currently active tab is removed', async () => {
+			it( 'should not have any selected tabs if the currently selected tab is removed, even if a tab is matching the defaultTabId', async () => {
```
Comment (ciampo) [inline packages/components/src/tabs/test/index.tsx:741]: This test was updated to test the new behavior

File: packages/components/src/tabs/test/index.tsx [hunk 11]
```diff
@@ -747,10 +763,20 @@ describe( 'Tabs', () => {
...
+				);
+
+				// No tabpanel should be rendered either
+				expect(
+					screen.queryByRole( 'tabpanel' )
+				).not.toBeInTheDocument();
 			} );
 
-			it( 'should fall back to the tab associated to `defaultTabId` if the currently active tab becomes disabled', async () => {
+			it( 'should keep the currently selected tab even if it becomes disabled', async () => {
```
Comment (ciampo) [inline packages/components/src/tabs/test/index.tsx:779]: This test was updated to test the new behavior

File: packages/components/src/tabs/test/index.tsx [hunk 12]
```diff
@@ -918,7 +956,7 @@ describe( 'Tabs', () => {
 				expect( await getSelectedTab() ).toHaveTextContent( 'Beta' );
 			} );
 
-			it( 'should select first enabled tab when the tab associated to `defaultTabId` is disabled', async () => {
+			it( 'should select the tab associated to `defaultTabId` even if the tab is disabled', async () => {
```
Comment (ciampo) [inline packages/components/src/tabs/test/index.tsx:959]: This test was updated to test the new behavior

File: packages/components/src/tabs/test/index.tsx [hunk 13]
```diff
@@ -948,10 +986,10 @@ describe( 'Tabs', () => {
 
 				// Even if the initial tab becomes enabled again, the selected tab doesn't
 				// change.
-				expect( await getSelectedTab() ).toHaveTextContent( 'Gamma' );
+				expect( await getSelectedTab() ).toHaveTextContent( 'Beta' );
 			} );
 
-			it( 'should select the first enabled tab when the selected tab becomes disabled', async () => {
+			it( 'should keep the currently tab as selected even when it becomes disabled', async () => {
```
Comment (ciampo) [inline packages/components/src/tabs/test/index.tsx:992]: This test was updated to test the new behavior

File: packages/components/src/tabs/test/index.tsx [hunk 14]
```diff
@@ -981,21 +1019,19 @@ describe( 'Tabs', () => {
...
 
-				expect( await getSelectedTab() ).toHaveTextContent( 'Beta' );
-				expect( mockOnSelect ).toHaveBeenCalledTimes( 2 );
-				expect( mockOnSelect ).toHaveBeenLastCalledWith( 'beta' );
+				expect( await getSelectedTab() ).toHaveTextContent( 'Alpha' );
+				expect( mockOnSelect ).toHaveBeenCalledTimes( 1 );
 			} );
 
-			it( 'should select the first enabled tab when the tab associated to `defaultTabId` becomes disabled while being the active tab', async () => {
+			it( 'should select the tab associated to `defaultTabId` even when disabled', async () => {
```
Comment (ciampo) [inline packages/components/src/tabs/test/index.tsx:1034]: This test was updated to test the new behavior

File: packages/components/src/tabs/test/index.tsx [hunk 15]
```diff
@@ -1072,7 +1106,7 @@ describe( 'Tabs', () => {
 
 			expect( await getSelectedTab() ).toHaveTextContent( 'Gamma' );
 		} );
-		it( 'should not render any tab if `selectedTabId` does not match any known tab', async () => {
+		it( 'should not have a selected tab if `selectedTabId` does not match any known tab', async () => {
```
Comment (ciampo) [inline packages/components/src/tabs/test/index.tsx:1109]: This test was update to reflect the new behaviour

File: packages/components/src/tabs/test/index.tsx [hunk 16]
```diff
@@ -1087,11 +1121,13 @@ describe( 'Tabs', () => {
 			// No tabpanel should be rendered either
 			expect( screen.queryByRole( 'tabpanel' ) ).not.toBeInTheDocument();
 		} );
-		it( 'should not render any tab if the active tab is removed', async () => {
+		it( 'should not have a selected tab if the active tab is removed, but should select a tab that gets added if it matches the selectedTabId', async () => {
```
Comment (ciampo) [inline packages/components/src/tabs/test/index.tsx:1124]: This test was updated to test the new behavior

File: packages/components/src/tabs/test/index.tsx [hunk 17]
```diff
@@ -1118,17 +1155,11 @@ describe( 'Tabs', () => {
...
-				screen.queryByRole( 'tab', { selected: true } )
-			).not.toBeInTheDocument();
-			// No tabpanel should be rendered either
-			expect( screen.queryByRole( 'tabpanel' ) ).not.toBeInTheDocument();
+			expect( await getSelectedTab() ).toHaveTextContent( 'Beta' );
 		} );
 
 		describe( 'Disabled tab', () => {
-			it( 'should not render any tab if `selectedTabId` refers to a disabled tab', async () => {
+			it( 'should `selectedTabId` refers to a disabled tab', async () => {
```
Comment (ciampo) [inline packages/components/src/tabs/test/index.tsx:1162]: This test was updated to test the new behavior

File: packages/components/src/tabs/test/index.tsx [hunk 18]
```diff
@@ -1149,18 +1180,9 @@ describe( 'Tabs', () => {
...
-					).not.toBeInTheDocument();
-				} );
-				// No tabpanel should be rendered either
-				expect(
-					screen.queryByRole( 'tabpanel' )
-				).not.toBeInTheDocument();
+				expect( await getSelectedTab() ).toHaveTextContent( 'Beta' );
 			} );
-			it( 'should not render any tab when the selected tab becomes disabled', async () => {
+			it( 'should keep the currently selected tab as selected even when it becomes disabled', async () => {
```
Comment (ciampo) [inline packages/components/src/tabs/test/index.tsx:1185]: This test was updated to test the new behavior

**Comment by ciampo** [comment]
@tyxla I've added more details on the behavioral changes added by this PR. Hopefully, they can help in testing such changes (even if, probably, most of the behavioral changes can't be currently experienced in the editor). 

Let me know if you'd like more details — otherwise, this PR is ready for a review round.

**Comment by tyxla** [comment]
@ciampo sorry - I couldn't make it to testing and reviewing this one today, but planning to devote some time tomorrow.

File: packages/components/src/tabs/tab.tsx [hunk 19]
```diff
@@ -22,20 +27,35 @@ export const Tab = forwardRef<
...
 >( function Tab( { children, tabId, disabled, render, ...otherProps }, ref ) {
-	const context = useTabsContext();
-	if ( ! context ) {
+	const { store, instanceId } = useTabsContext() ?? {};
+
+	// If the active item is not connected, the tablist may end up in a state
+	// where none of the tabs are tabbable. In this case, we force all tabs to
+	// be tabbable, so that as soon as an item received focus, it becomes active
+	// and Tablist goes back to working as expected.
+	// eslint-disable-next-line @wordpress/no-unused-vars-before-return
```
Comment (tyxla) [inline packages/components/src/tabs/tab.tsx:36]: We might want to refactor those early returns in the future. Do we have an open issue for that?

**Review by tyxla — APPROVED** [review]
LGTM 👍 

I've double-checked the custom behaviors and didn't spot anything unexpected.

Thanks @ciampo, this is a great cleanup 🚀

File: packages/components/src/tabs/tab.tsx (hunk 19 above)
Comment (ciampo) [reply packages/components/src/tabs/tab.tsx:36]: Do you mean the "Subcomponent must be wrapped in the top level component" warnings?  Just double-checking before I go ahead and create the issue.

Comment (tyxla) [reply packages/components/src/tabs/tab.tsx:36]: Yes 👍

Comment (ciampo) [reply packages/components/src/tabs/tab.tsx:36]: https://github.com/WordPress/gutenberg/issues/66530

**Comment by tyxla** [comment]
The performance impact of this PR on [inserter opening](https://www.codevitals.run/project/gutenberg/inserterOpen) is impressive:

Screenshot 2024-10-29 at 11 16 59

**Comment by ciampo** [comment]
A good reminder that we should aim for simplicity when possible. Low-level components can have many instances rendered at once, and even a small improvement can make a big difference in the metrics 🥳

**Comment by ellatrix** [comment]
@ciampo Curious, what has changed here? Just rendering less components?

**Comment by ciampo** [comment]
@ellatrix  we removed a few hooks that were applying some logic to `Tabs`'s initial selection and re-selection if the selected tab becomes disabled or is removed (see https://github.com/WordPress/gutenberg/issues/66011 for more details).

The tabs in the inserter sidebar used some (but not all) of that logic, and so we only re-added the necessary bits (see https://github.com/WordPress/gutenberg/pull/66097#discussion_r1801253430)
